File: src/api/session-information.ts
import { defaultFetcher, type Fetcher, type Request, setCookiesArrayToRequest } from "@literate.ink/utilities";
import { decodeSessionInformation } from "~/decoders/session-information";
import { encodeAccountKindToPath } from "~/encoders/account-kind";
import { BusyPageError, PageUnavailableError, SuspendedIPError, type AccountKind, type SessionInformation } from "~/models";

export const sessionInformation = async (options: {
  base: string
  kind: AccountKind,
  params: Record<string, any>,
  cookies: string[]
}, fetcher: Fetcher = defaultFetcher): Promise<SessionInformation> => {
  const url = new URL(options.base + "/" + encodeAccountKindToPath(options.kind));
  for (const [key, value] of Object.entries(options.params)) {
    url.searchParams.set(key, value);
  }

  const request: Request = { url, redirect: "manual" };
  setCookiesArrayToRequest(request, options.cookies);

  const { content: html } = await fetcher(request);

  try {
    // Remove all spaces and line breaks.
    const body_cleaned = html.replace(/ /ug, "").replace(/\n/ug, "");

    // Find the relaxed JSON of the session.
    const from = "Start(";
    const to = ")}catch";

    const relaxed_data = body_cleaned.substring(
      body_cleaned.indexOf(from) + from.length,
      body_cleaned.indexOf(to)
    );

    // Convert the relaxed JSON to something we can parse.
    const session_data_string = relaxed_data
      .replace(/(['"])?([a-z0-9A-Z_]+)(['"])?:/gu, "\"$2\": ")
      .replace(/'/gu, "\"");

    return decodeSessionInformation(JSON.parse(session_data_string), options.base);
  }
  catch (error) {
    if (html.includes("Votre adresse IP est provisoirement suspendue")) {
      throw new SuspendedIPError();
    }
    else if (html.includes("Le site n'est pas disponible")) {
      throw new PageUnavailableError();
    }
    else if (html.includes("Le site est momentanément indisponible")) {
      throw new BusyPageError();
    }

    throw new PageUnavailableError();
  }
};

File: library/src/commonMain/kotlin/api/session-information.kt
package api

import decoders.decodeSessionInformation
import encoders.encodeAccountKindToPath
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import models.AccountKind
import models.SessionInformation
import models.errors.BusyPageError
import models.errors.PageUnavailableError
import models.errors.SuspendedIPError

data class SessionInfoParams(
    val base: String,
    val kind: AccountKind,
    val params: Map<String, String>,
    val cookies: List<String>
)

suspend fun sessionInformation(
    options: SessionInfoParams,
    httpClient: HttpClient = HttpClient {followRedirects = false}
): SessionInformation {
    val url = Url(options.base + "/" + encodeAccountKindToPath(options.kind))

    val response = httpClient.get(url) {
        url {
            options.params.forEach { (k, v) -> parameters.append(k, v) }
        }
    }

    val content = response.bodyAsText()

    try {
        val bodyCleaned = content.replace(" ", "").replace("\n", "")
        val from = "Start("
        val to = ")}catch"

        val relaxedData = bodyCleaned.substring(
            bodyCleaned.indexOf(from) + from.length,
            bodyCleaned.indexOf(to)
        )

        val sessionDataString = relaxedData
            .replace(Regex("(['\"])?([a-z0-9A-Z_]+)(['\"])?:"), "\"$2\": ")
            .replace(Regex("'"), "\"")

        return decodeSessionInformation(Json.parseToJsonElement(sessionDataString).jsonObject, options.base)
    }
    catch (err: Exception) {
        if (content.contains("Votre adresse IP est provisoirement suspendue"))
            throw SuspendedIPError()
        else if (content.contains("Le site n'est pas disponible"))
            throw PageUnavailableError()
        else if (content.contains("Le site est momentanément indisponible"))
            throw BusyPageError()

        throw PageUnavailableError()
    }
}